File: app/routers/traffic.py
"""Endpoints para consumo de ancho de banda por usuario."""

import os
from datetime import datetime, timezone
from fastapi import APIRouter, HTTPException, Header, Query
from pydantic import BaseModel
from typing import Optional, List
from app.database import get_db
from app.security import decode_token

router = APIRouter(prefix="/traffic", tags=["traffic"])

SYNC_KEY = os.environ.get("NAC_SYNC_KEY", "nac-sync-2024")

# Cache en memoria de la última sesión reportada por el router (se refresca cada 60s)
_latest_sessions: dict[str, dict] = {}
_last_report_at: Optional[str] = None


async def verify_token(authorization: Optional[str]):
    if not authorization or not authorization.startswith("Bearer "):
        raise HTTPException(status_code=401, detail="No autenticado")
    payload = decode_token(authorization.split(" ")[1])
    if not payload:
        raise HTTPException(status_code=401, detail="Token inválido")
    return payload


def _fmt_bytes(b: int) -> dict:
    if b >= 1_073_741_824:
        return {"value": round(b / 1_073_741_824, 2), "unit": "GB"}
    if b >= 1_048_576:
        return {"value": round(b / 1_048_576, 2), "unit": "MB"}
    if b >= 1024:
        return {"value": round(b / 1024, 2), "unit": "KB"}
    return {"value": b, "unit": "B"}


class SessionReport(BaseModel):
    user: str
    address: str = ""
    mac_address: str = ""
    uptime: str = ""
    bytes_in: int = 0
    bytes_out: int = 0


class TrafficReportRequest(BaseModel):
    sessions: List[SessionReport]


@router.post("/report")
async def receive_traffic_report(
    body: TrafficReportRequest,
    x_sync_key: Optional[str] = Header(None),
):
    """
    Endpoint que el router MikroTik llama para reportar sesiones activas.
    Actualiza la tabla traffic_usage con el consumo de cada usuario.
    """
    if x_sync_key != SYNC_KEY:
        raise HTTPException(status_code=403, detail="Clave de sincronización inválida")

    db = await get_db()
    try:
        saved = 0
        for s in body.sessions:
            if s.bytes_in == 0 and s.bytes_out == 0:
                continue

            # El router envía la MAC como campo 'user' y también en 'mac_address'
            mac = (s.mac_address or s.user or "").upper().strip()
            if not mac:
                continue

            # Buscar usuario por MAC en la tabla devices
            cur = await db.execute(
                "SELECT user_id FROM devices WHERE UPPER(mac_address) = ?", (mac,)
            )
            device_row = await cur.fetchone()
            if not device_row:
                continue

            user_id = device_row[0]

            await db.execute(
                "INSERT INTO traffic_usage (user_id, bytes_down, bytes_up) VALUES (?, ?, ?)",
                (user_id, s.bytes_in, s.bytes_out)
            )
            await db.execute(
                "UPDATE users SET last_login = CURRENT_TIMESTAMP WHERE id = ?",
                (user_id,)
            )
            saved += 1

        await db.commit()

        # Actualizar cache en memoria con las sesiones actuales del router
        global _latest_sessions, _last_report_at
        _latest_sessions = {
            s.mac_address.upper(): {
                "user":        s.user,
                "address":     s.address,
                "mac_address": s.mac_address,
                "uptime":      s.uptime,
                "bytes_in":    s.bytes_in,
                "bytes_out":   s.bytes_out,
            }
            for s in body.sessions
            if s.mac_address
        }
        _last_report_at = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S UTC")

        return {"ok": True, "saved": saved}
    finally:
        await db.close()


@router.get("/live")
async def get_live_sessions(key: str = Query(default="")):
    """
    Sesiones activas en tiempo real tal como el router las reportó (última vez).
    Autenticado con sync_key — no requiere login de admin.
    """
    if key != SYNC_KEY:
        raise HTTPException(status_code=403, detail="Clave inválida")

    sessions = list(_latest_sessions.values())
    return {
        "last_report_at": _last_report_at,
        "session_count":  len(sessions),
        "sessions":       sessions,
    }


@router.get("/users")
async def get_traffic_per_user(authorization: Optional[str] = Header(None)):
    """Retorna consumo acumulado de ancho de banda por usuario."""
    payload = await verify_token(authorization)

    db = await get_db()
    try:
        cur = await db.execute(
            "SELECT role FROM users WHERE username = ?", (payload.get("sub"),)
        )
        caller = await cur.fetchone()
        if not caller or caller[0] != "admin":
            raise HTTPException(status_code=403, detail="Solo administradores")

        # Sesiones activas reportadas recientemente (últimos 5 minutos)
        cur = await db.execute("""
            SELECT u.id, u.username, u.full_name,
                   COALESCE(SUM(t.bytes_down), 0) as total_down,
                   COALESCE(SUM(t.bytes_up), 0)   as total_up,
                   MAX(t.recorded_at) as last_seen
            FROM users u
            LEFT JOIN traffic_usage t ON t.user_id = u.id
            WHERE u.role = 'user' AND u.approval_status = 'approved'
            GROUP BY u.id
            ORDER BY (COALESCE(SUM(t.bytes_down),0) + COALESCE(SUM(t.bytes_up),0)) DESC
        """)
        rows = await cur.fetchall()

        # Usuarios vistos en los últimos 10 minutos = "online"
        cur_online = await db.execute("""
            SELECT DISTINCT user_id FROM traffic_usage
            WHERE recorded_at >= datetime('now', '-10 minutes')
        """)
        online_ids = {r[0] for r in await cur_online.fetchall()}

        result = []
        for row in rows:
            uid, uname, fname, total_down, total_up, last_seen = row
            total_down = total_down or 0
            total_up   = total_up   or 0
            result.append({
                "user_id":    uid,
                "username":   uname,
                "full_name":  fname,
                "online":     uid in online_ids,
                "bytes_down": total_down,
                "bytes_up":   total_up,
                "bytes_total": total_down + total_up,
                "down_fmt":   _fmt_bytes(total_down),
                "up_fmt":     _fmt_bytes(total_up),
                "total_fmt":  _fmt_bytes(total_down + total_up),
                "last_seen":  last_seen,
            })

        return result
    finally:
        await db.close()

File: php/api/traffic.php
<?php
// /api/traffic/*  — port de app/routers/traffic.py

function traffic_report() {
    if (header_val('x-sync-key') !== SYNC_KEY) abort(403, 'Clave de sincronización inválida');
    $b = body_json();
    $sessions = $b['sessions'] ?? [];
    $saved = 0;
    $db = db();
    $db->beginTransaction();
    try {
        // refrescar cache de sesiones en vivo
        $db->exec("DELETE FROM live_sessions");
        foreach ($sessions as $s) {
            // Contadores acumulados del router (/ip hotspot host):
            //   bytes-in  = recibido DEL cliente  = SUBIDA del cliente
            //   bytes-out = enviado AL cliente    = BAJADA del cliente
            $rin  = max(0, (int)($s['bytes_in'] ?? 0));   // subida acumulada
            $rout = max(0, (int)($s['bytes_out'] ?? 0));  // bajada acumulada
            $mac = strtoupper(trim($s['mac_address'] ?? ($s['user'] ?? '')));
            if ($mac !== '') {
                q("INSERT OR REPLACE INTO live_sessions (mac_address, user, address, uptime, bytes_in, bytes_out, reported_at)
                   VALUES (?, ?, ?, ?, ?, ?, CURRENT_TIMESTAMP)",
                   [$mac, $s['user'] ?? '', $s['address'] ?? '', $s['uptime'] ?? '', $rin, $rout]);
            }
            if ($mac === '') continue;
            $dev = q_one("SELECT user_id FROM devices WHERE UPPER(mac_address) = ?", [$mac]);
            if (!$dev) continue;

            // Delta real desde el último reporte; si el contador bajó (host recreado), el delta es el valor actual.
            $prev = q_one("SELECT last_in, last_out FROM mac_counter WHERE mac_address = ?", [$mac]);
            $lin  = $prev ? (int)$prev['last_in']  : 0;
            $lout = $prev ? (int)$prev['last_out'] : 0;
            $dUp   = ($rin  >= $lin)  ? ($rin  - $lin)  : $rin;   // subida incremental
            $dDown = ($rout >= $lout) ? ($rout - $lout) : $rout;  // bajada incremental

            q("INSERT OR REPLACE INTO mac_counter (mac_address, last_in, last_out, updated_at)
               VALUES (?, ?, ?, CURRENT_TIMESTAMP)", [$mac, $rin, $rout]);

            if ($dUp > 0 || $dDown > 0) {
                // bytes_down = bajada, bytes_up = subida (mapeo corregido)
                q("INSERT INTO traffic_usage (user_id, bytes_down, bytes_up) VALUES (?, ?, ?)",
                  [$dev['user_id'], $dDown, $dUp]);
                q("UPDATE users SET last_login = CURRENT_TIMESTAMP WHERE id = ?", [$dev['user_id']]);
                $saved++;
            }
        }
        $db->commit();
    } catch (Throwable $e) {
        $db->rollBack();
        abort(500, 'report error: ' . $e->getMessage());
    }
    json_out(['ok' => true, 'saved' => $saved]);
}

function traffic_live() {
    if (($_GET['key'] ?? '') !== SYNC_KEY) abort(403, 'Clave inválida');
    $rows = q_all("SELECT user, address, mac_address, uptime, bytes_in, bytes_out FROM live_sessions");
    $last = q_one("SELECT MAX(reported_at) AS t FROM live_sessions");
    $sessions = [];
    foreach ($rows as $r) {
        $sessions[] = ['user' => $r['user'], 'address' => $r['address'],
                       'mac_address' => $r['mac_address'], 'uptime' => $r['uptime'],
                       'bytes_in' => (int)$r['bytes_in'], 'bytes_out' => (int)$r['bytes_out']];
    }
    json_out(['last_report_at' => $last['t'] ?? null,
              'session_count' => count($sessions), 'sessions' => $sessions]);
}

function _valid_date($d, $default) {
    return (is_string($d) && preg_match('/^\d{4}-\d{2}-\d{2}$/', $d)) ? $d : $default;
}

// GET /api/traffic/report-range?from=YYYY-MM-DD&to=YYYY-MM-DD[&user_id=N]
// Reporte de consumo por empleado en un rango de fechas. Solo admin.
function traffic_report_range() {
    require_admin();
    $to   = _valid_date($_GET['to']   ?? null, gmdate('Y-m-d'));
    $from = _valid_date($_GET['from'] ?? null, gmdate('Y-m-d', time() - 29 * 86400));
    if ($from > $to) { $tmp = $from; $from = $to; $to = $tmp; }
    $user_id = isset($_GET['user_id']) && $_GET['user_id'] !== '' ? (int)$_GET['user_id'] : null;

    $params = [$from, $to];
    $userFilter = '';
    if ($user_id) { $userFilter = ' AND u.id = ?'; $params[] = $user_id; }

    $rows = q_all(
        "SELECT u.id, u.username, u.full_name, u.department, u.company, u.role,
                COUNT(t.id) AS sessions,
                COALESCE(SUM(t.bytes_down),0) AS d,
                COALESCE(SUM(t.bytes_up),0)   AS u_up,
                MAX(t.recorded_at) AS last_seen
         FROM users u
         JOIN traffic_usage t ON t.user_id = u.id
         WHERE date(t.recorded_at) BETWEEN ? AND ?$userFilter
         GROUP BY u.id
         ORDER BY (COALESCE(SUM(t.bytes_down),0) + COALESCE(SUM(t.bytes_up),0)) DESC",
        $params
    );

    $users = [];
    $tDown = 0; $tUp = 0;
    foreach ($rows as $r) {
        $down = (int)$r['d']; $up = (int)$r['u_up'];
        $tDown += $down; $tUp += $up;
        $users[] = [
            'user_id' => (int)$r['id'], 'username' => $r['username'], 'full_name' => $r['full_name'],
            'department' => $r['department'], 'company' => $r['company'], 'role' => $r['role'],
            'sessions' => (int)$r['sessions'],
            'bytes_down' => $down, 'bytes_up' => $up, 'bytes_total' => $down + $up,
            'down_fmt' => fmt_bytes($down), 'up_fmt' => fmt_bytes($up), 'total_fmt' => fmt_bytes($down + $up),
            'last_seen' => $r['last_seen'],
        ];
    }

    // Desglose diario (solo si se filtró un empleado)
    $daily = [];
    if ($user_id) {
        $drows = q_all(
            "SELECT date(recorded_at) AS day,
                    COALESCE(SUM(bytes_down),0) AS d, COALESCE(SUM(bytes_up),0) AS u_up
             FROM traffic_usage
             WHERE user_id = ? AND date(recorded_at) BETWEEN ? AND ?
             GROUP BY day ORDER BY day",
            [$user_id, $from, $to]
        );
        foreach ($drows as $r) {
            $down = (int)$r['d']; $up = (int)$r['u_up'];
            $daily[] = [
                'date' => $r['day'], 'bytes_down' => $down, 'bytes_up' => $up, 'bytes_total' => $down + $up,
                'down_fmt' => fmt_bytes($down), 'up_fmt' => fmt_bytes($up), 'total_fmt' => fmt_bytes($down + $up),
            ];
        }
    }

    json_out([
        'from' => $from, 'to' => $to, 'user_id' => $user_id,
        'users' => $users,
        'totals' => [
            'bytes_down' => $tDown, 'bytes_up' => $tUp, 'bytes_total' => $tDown + $tUp,
            'down_fmt' => fmt_bytes($tDown), 'up_fmt' => fmt_bytes($tUp), 'total_fmt' => fmt_bytes($tDown + $tUp),
        ],
        'daily' => $daily,
    ]);
}

function traffic_users() {
    require_admin();
    $rows = q_all(
        "SELECT u.id, u.username, u.full_name, u.role,
                COALESCE(SUM(t.bytes_down),0) AS total_down,
                COALESCE(SUM(t.bytes_up),0)   AS total_up,
                MAX(t.recorded_at) AS last_seen
         FROM users u
         LEFT JOIN traffic_usage t ON t.user_id = u.id
         WHERE u.approval_status = 'approved'
           AND EXISTS (SELECT 1 FROM devices d WHERE d.user_id = u.id)
         GROUP BY u.id
         ORDER BY (COALESCE(SUM(t.bytes_down),0) + COALESCE(SUM(t.bytes_up),0)) DESC"
    );
    $online = q_all("SELECT DISTINCT user_id FROM traffic_usage WHERE recorded_at >= datetime('now','-10 minutes')");
    $online_ids = array_map(fn($r) => (int)$r['user_id'], $online);

    $out = [];
    foreach ($rows as $r) {
        $down = (int)$r['total_down']; $up = (int)$r['total_up'];
        $out[] = [
            'user_id' => (int)$r['id'], 'username' => $r['username'], 'full_name' => $r['full_name'],
            'role' => $r['role'],
            'online' => in_array((int)$r['id'], $online_ids, true),
            'bytes_down' => $down, 'bytes_up' => $up, 'bytes_total' => $down + $up,
            'down_fmt' => fmt_bytes($down), 'up_fmt' => fmt_bytes($up),
            'total_fmt' => fmt_bytes($down + $up), 'last_seen' => $r['last_seen'],
        ];
    }
    json_out($out);
}
